pio/file_linux.go

```diff
@@ -0,0 +1,283 @@
+		}
+		for i := range ioctx.res {
+			ioctx.res[i] = io_event{}
+		}
+		ioctx.ptr = ioctx.ptr[:0]
+		ioctx.req = ioctx.req[:0]
+		ioctx.res = ioctx.res[:0]
+		ioctxPool <- ioctx
+	}()
+
+	fd := f.Fd()
+	for i := range ops {
+		op := &ops[i]
+		ioctx.res = append(ioctx.res, io_event{})
+		ioctx.req = append(ioctx.req, iocb{
+			data:    uint64(i),
+			opcode:  pread,
+			filedes: uint32(fd),
+			buf:     op.data(),
+			nbytes:  op.size(),
+			offset:  op.Off,
+		})
+		ioctx.ptr = append(ioctx.ptr, &ioctx.req[i])
+	}
+
+	pending := 0
+	for {
+		if len(ioctx.ptr) > 0 {
+			n, errno := io_submit(ioctx.ctx, ioctx.ptr)
+			if errno != 0 {
+				// When resubmitting the operations fail, we can afford to
+				// only abort these operations and let the other complete.
+				err := os.NewSyscallError("io_submit", errno)
+
+				for _, p := range ioctx.ptr {
+					ops[p.data].Err = err
+				}
+			} else {
+				pending += n
+				ioctx.ptr = ioctx.ptr[:copy(ioctx.ptr, ioctx.ptr[n:])]
+			}
+		}
+
+		if pending == 0 {
+			return
+		}
+
+		n, errno := io_getevents(ioctx.ctx, 1, ioctx.res)
+		switch errno {
+		case 0:
+			for i := range ioctx.res[:n] {
+				r := &ioctx.res[i]
+				op := &ops[r.data]
+				rn := int(r.res) + byteDist(op.data(), ioctx.req[r.data].buf)
+
+				// According to https://lkml.iu.edu/hypermail/linux/kernel/0304.3/1296.html
+				// the `io_event.res` field will be negative value holding the
+				// error code on failure.
+				if r.res < 0 {
+					op.Err = os.NewSyscallError("pread", syscall.Errno(-r.res))
+				} else if rn < len(op.Data) {
+					op.Err = io.EOF
+					op.Data = op.Data[:rn]
+				}
+
+				pending--
+			}
+
+		case syscall.EINTR:
+			// If the syscall is interrupted by a signal, some operations may
+			// have only partially completed. Resubmit.
+			for i := range ioctx.res[:n] {
+				r := &ioctx.res[i]
+				p := &ioctx.req[r.data]
+				op := &ops[r.data]
+				rn := int(r.res) + byteDist(op.data(), p.buf)
+
+				if r.res < 0 {
+					op.Err = os.NewSyscallError("pread", syscall.Errno(-r.res))
+				} else if rn < len(op.Data) {
+					// We don't know whether we reached the end of the file or
+					// the read was interrupted by a signal handler, so we must
+					// resubmit the operation until io_getevents returns
+					// successfully.
+					//
+					// We adjust the buffer's base pointer, offset and length
+					// to account for the data that may have already been read.
+					p.buf = byteAdd(p.buf, int(r.res))
+					p.nbytes -= uint64(r.res)
+					p.offset += int64(r.res)
+					ioctx.ptr = append(ioctx.ptr, p)
+					continue
+				}
+
+				pending--
+			}
+
+		default:
+			err := os.NewSyscallError("io_getevents", errno)
+
+			for i := range ops {
+				op := &ops[i]
+				op.Data, op.Err = op.Data[:0], err
+			}
+
+			return
+		}
+	}
+}
+
+type ioctx struct {
+	ctx io_context_t
+	ptr []*iocb
+	req []iocb
+	res []io_event
+}
+
+const (
+	// Hard limits, chosen arbitrarily, we should revisit if they are not
```

Comment:
Should we make those configuration so that we can experiment with those values without recompiling things ?